### DISCUSSION TITLE
ImportError: cannot import name 'Kinetics400' from 'torchvision.datasets'

### DISCUSSION THREAD

**Discussion by wjkbfq** [yoshitomo-matsubara/torchdistill#282] [Q&A]
Dear author,
when run `!python torchdistill/examples/image_classification.py --config torchdistill/configs/sample/cifar10/ce/resnet20-hyperparameter_tuning.yaml --log log/cifar10/ce/resnet20-hyperparameter_tuning.log` in colab, I got the following error report.

```
Traceback (most recent call last):
  File "torchdistill/examples/image_classification.py", line 15, in <module>
    from torchdistill.core.distillation import get_distillation_box
  File "/usr/local/lib/python3.8/dist-packages/torchdistill/core/distillation.py", line 15, in <module>
    from torchdistill.datasets.util import build_data_loaders
  File "/usr/local/lib/python3.8/dist-packages/torchdistill/datasets/util.py", line 8, in <module>
    from torchvision.datasets import PhotoTour, Kinetics400, HMDB51, UCF101, Cityscapes, CocoCaptions, CocoDetection, \
ImportError: cannot import name 'Kinetics400' from 'torchvision.datasets' (/usr/local/lib/python3.8/dist-packages/torchvision/datasets/__init__.py)

```

**Comment by yoshitomo-matsubara** [comment]
Hi @wjkbfq 

Could you share your package versions of
1. torch
2. torchvision
3. torchdistill (on top of that, is it locally installed or PyPI package)

?

If you're using torchvision > v0.13.1, I'd suggest you use torchvision=<v0.13.1 instead

**Reply by AhmedHussKhalifa** [reply]
I have used the following but still the same problem. 

pip install torch==1.12.0+cu116 torchvision==0.13.0+cu116 torchaudio==0.12.0 --extra-index-url https://download.pytorch.org/whl/cu116

**Reply by yoshitomo-matsubara** [reply]
Hi @AhmedHussKhalifa 

Please make sure `torchvision.__version__` returns <= 0.13.1
```python
import torch
import torchvision
print(torch.__version__)
print(torchvision.__version__)
```

I think specifying versions in pip doesn't always install the right versions due to dependencies

**Reply by AhmedHussKhalifa** [reply]
The main problem is  here:
DATASET_DICT.update(misc_util.get_classes_as_dict('torchvision.datasets'))

We can change it:

dataset_cls_or_func = DATASET_DICT[dataset_type.lower()]

--- 

To answer your question:
print(torch.__version__)
1.12.1
print(torchvision.__version__)
0.13.1

**Reply by yoshitomo-matsubara** [reply]
Hi @AhmedHussKhalifa 

I don't think
`DATASET_DICT.update(misc_util.get_classes_as_dict('torchvision.datasets'))` 
is the problem and your change won't resolve the error because it's just getting all the classes under torchvision.datasets to be registered as part of `DATASET_DICT`.

But as the error message in the original question shows, 
https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/datasets/util.py#L8
is the problem since it specifies `Kinetics400` class while it no longer exists from 0.14.0, and I confirmed torchvision=0.13.1+cu102 on my end works without the error.
The error is being resolved locally and the change will be reflected to the next release.

**Reply by AhmedHussKhalifa** [reply]
I have installed trochvision --> 0.13.1+cu113 as my GPU does not support cu102. I believe this will make a huge difference. However, the same error is still present. 

<img width="890" alt="image" src="https://user-images.githubusercontent.com/9024871/218891364-b6f8bc43-ab6c-4468-93c0-383692bcbef7.png">

**Reply by AhmedHussKhalifa** [reply]
I have managed to install torchvision 0.13.1+cu102. However, the same error is still present.
<img width="1017" alt="image" src="https://user-images.githubusercontent.com/9024871/218897399-edf71f17-7b78-4a8b-9fed-580a5d6ef652.png">

**Reply by yoshitomo-matsubara** [reply]
Hi @AhmedHussKhalifa 

From the file paths in your screenshot, I suspect 
1. cu102 vs. cu113 is not making the error, and 
2. you're using locally installed torchditill instead of its PyPI package ver.

Uninstall the locally installed torchdistill and reinstall it by `pip3 install torchdistill` as suggested [here](https://github.com/yoshitomo-matsubara/torchdistill/discussions/278#discussioncomment-4700064).
I am making a lot of changes for next release that are not synced with example scripts yet

**Reply by AhmedHussKhalifa** [reply]
Could you provide an example if I want to modify the loss function for the KD?

I used to modify this:

https://github.com/yoshitomo-matsubara/torchdistill/blob/a6454f4c563ba8e3d6789feb193ea76bdd10627c/torchdistill/losses/single.py#L52

**Reply by AhmedHussKhalifa** [reply]
If I want to register a new KDloss function, Could I do the following?
I have followed the example that you have provided in some sense.
Let's do that as an example:

<img width="864" alt="image" src="https://user-images.githubusercontent.com/9024871/218908699-cf347d2b-5496-4b00-875f-9d81da1c7751.png">

**Reply by yoshitomo-matsubara** [reply]
Please check the comment shared above carefully. It includes a suggested way (step 3).

Also, in general, please use this space more respectfully.
I answered [all your questions in Discussions](https://github.com/yoshitomo-matsubara/torchdistill/discussions?discussions_q=AhmedHussKhalifa), but you closed only one of the four discussions (did not either respond to or choose my answers for the other three discussions) and repeated questions I've already answered for other people.
